0004-market-integration-token-standard.md

```diff
@@ -0,0 +1,196 @@
+- I want to see the rarity of a token (total supply)
+
+## Example of minting a token
+
+```TypeScript
+let tokenTypeId = mintTokenType("Corgi Token", 1000);
+near.log(tokenTypeId);
+// 0
+
+// This should be done with a constructor, but is done here with attributes illustratively
+let corgiToken = {};
+corgiToken.name = "Winnifred";
+corgiToken.tokenTypeId = tokenTypeId;
+corgiToken.data = "{\"url\":\"corgi name\"}";
+
+// At this point, corgi doesn't have an id because it hasn't been minted yet
+near.log(corgiToken.id);
+// null
+
+// ID generation should be incremental and unique across all tokens
+let mintedCorgiToken = mintToken(corgiToken);
+near.log(corgiToken.id)
+// 0
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Template for smart contract in AssemblyScript
+
+At time of writing, this standard is established with several constraints found in AssemblyScript. The first is that interfaces are not an implemented feature of AssemblyScript, and the second is that classes are not exported in the conversion from AssemblyScript to WASM. This means that the entire contract could be implemented as a class, which might be better for code organization, but it would be deceiving in function.
+
+```TypeScript
+  type TokenTypeId = u32;
+  type TokenId = u64;
+
+  function unimplemented():void {
+    near.log("Function not implemented");
+  }
+
+  export function init():void {
+    unimplemented();
+  }
+
+  export function mintTokenType(name:string, totalSupply:u64):TokenTypeId {
+    unimplemented();
+  }
+
+  export function mintToken(ownerId:string, token:Token):Token {
+    unimplemented();
+  }
+
+  export function getToken(tokenId:TokenId):Token {
+    unimplemented();
+  }
+
+  export function getTokenType(tokenTypeId:TokenTypeId):TokenType {
+    unimplemented();
+  }
+
+  export function getTokenOwner(tokenId:TokenId):string {
+    unimplemented();
+  }
+
+  export function getAllTokenTypes():TokenType[] {
+    unimplemented();
+  }
+
+  export function getTokensByOwner(ownerId:string, startTokenIdx:u64 = 0, limit:u32 = 10):Token[] {
+    unimplemented();
+  }
+
+  export function getCountByOwner(ownerId: string):u64 {
+    unimplemented();
+  }
+
+  export function transfer(ownerId:string, recipientId:string, tokenIds:TokenId[], onTransfer:PayloadCallback = null):void {
+    unimplemented();
+  }
+
+  class PayloadCallback {
+    contractId:string;
+    methodName:string;
+    payload: string;
+  }
+
+  class OnTransferArgs {
+    ownerId: string;
+    recipientId: string;
+    tokenIds:TokenId[];
+    payload: string;
+  }
+
+  //*** Researching the following methods ***//
+  function lockToken(tokenId:TokenId):void {
+    unimplemented();
+  }
+
+  function unlockToken(tokenId:TokenId):void {
+    unimplemented();
+  }
+
+```
+
+## Models
+
+```TypeScript
+export class Token {
+  id:u64;
+  tokenTypeId:TokenTypeId;
+  data:string;
```

Comment:
I think it would be helpful if there is more explanation on how the data field should be used,

Reply:
I'll add an example